lib/src/json_schema/json_schema.dart

```diff
@@ -425,36 +420,119 @@ class JsonSchema {
     // _logger.info('Completed Validating schema $_path'); TODO: re-add logger
   }
 
-  /// Given a path, find the ref'd [JsonSchema] from the map.
-  JsonSchema _getSchemaFromPath(String original) {
-    final String path = endPath(original);
-    JsonSchema result = _refMap[path];
+  /// Given a [Uri] path, find the ref'd [JsonSchema] from the map.
+  JsonSchema _getSchemaFromPath(Uri pathUri, [Set<Uri> refsEncountered]) {
+    // Store encountered refs to avoid cycles.
+    refsEncountered ??= {};
 
-    if (result == null) {
-      final schema = _freeFormMap[path];
-      if (schema is! Map) throw FormatExceptions.schema('free-form property $original at $path', schema);
-      return JsonSchema._fromMap(_root, schema, path);
+    Uri basePathUri;
+    if (pathUri.host.isEmpty && pathUri.path.isEmpty && (_uri != null || _inheritedUri != null)) {
+      // Prepend _uri or _inheritedUri to provided [Uri] path if no host or path detected.
+      pathUri = _translateLocalRefToFullUri(pathUri);
     }
 
-    // Follow the refs until a schema without a ref is reached.
-    final Set<String> _refsEncountered = Set<String>();
-    while (result.ref != null) {
-      if (!_refsEncountered.add(result.ref.toString())) {
-        throw FormatExceptions.error('Encountered ref cycle at ${result.ref}');
-      }
+    // basePathUri should always have an empty fragment.
+    basePathUri = Uri.parse('${pathUri.removeFragment()}#');
 
-      assert(endPath(result.ref.toString()) == result.ref.toString());
-      result = _refMap[result.ref.toString()];
+    // If _refMap already contains the full pathUri, return the ref'd JsonSchema.
+    if (_refMap.containsKey(pathUri.toString())) {
+      return _refMap[pathUri.toString()];
     }
 
-    return result;
+    JsonSchema baseSchema;
+    if (_refMap.containsKey(basePathUri.toString())) {
+      // Pull the baseSchema from the _refMap
+      baseSchema = _refMap[basePathUri.toString()];
+    } else {
+      // If the _refMap does not contain basePathUri, the ref was never resolved during validation.
+      throw ArgumentError('Unable to resolve path $pathUri');
```

Comment:
```suggestion
      throw ArgumentError('Failed to get schema for path because the schema file ($basePathUri) could not be found. Unable to resolve path $pathUri');
```

Reply:
Trying to make these errors a little more clear if we can, because similar errors have slowed me down in the past. Correct where I'm off base here?